### ISSUE TITLE
Improve repackage exclude to be able to specify a pattern

### ISSUE THREAD

**Issue by snicoll** [spring-projects/spring-boot#16563]
We've add a few reports around dependency exclusion for the repackage goal with Maven. The fact transitive artifacts are not filtered out is annoying and certain arrangements require a large number of artifacts to be defined.

While avoiding such practice in application's module should be preferred, it's not always possible. This issue is about reviewing our current support and investigating how we could improve the experience for this.

**Comment by JoshuaJeme** [comment]
Can we provide a spi which allow the user to add custom filters?

`org.springframework.boot.maven.FilterProvider` for example, has a method like `getFilters`

```java
public interface FilterProvider {
	List<ArtifactsFilter> getFilters();
}
```

User can provide a jar contain the Provider and config it through the dependency of maven plugin

```xml
<plugin>
  <groupId>org.springframework.boot</groupId>
  <artifactId>spring-boot-maven-plugin</artifactId>
  <version>2.2.0.M2</version>
  <dependencies>
    <dependency>
      <groupId>com.demo</groupId>
      <artifactId>filter-provider</artifactId>
      <version>0.0.1</version>
    </dependency>
  </dependencies>
</plugin>
```

**Comment by snicoll** [comment]
Thanks for the suggestion. That was one of the option when we brainstormed but we'd like to keep it as a fallback. The user still has to exclude transitive dependencies with more configuration and code to write. We'd like to research if we can't make exclude do that automatically instead.

**Comment by JoshuaJeme** [comment]
> Thanks for the suggestion. That was one of the option when we brainstormed but we'd like to keep it as a fallback. The user still has to exclude transitive dependencies with more configuration and code to write. We'd like to research if we can't make exclude do that automatically instead.

THX for the reply, I'm curious about what are the rest options, can you give more detail?

Another option I had thought is compound GAV pattern, like `groupId:artifactId:version:scope`, each part can use * as wildcard match (like https://maven.apache.org/docs/3.2.1/release-notes.html).

It's more flexable but still can't satisfy all the scenes, so I think the fallback solution is needed, which give the user a chance to add custom logic, it does has more configuration and code but better than no way.

**Comment by snicoll** [comment]
Yeah the wildcard thing was also discussed. What we'd like to research is a way to make exclude work the same way as the Maven model does. That way, you'd define the top-level artifact(s) you need to exclude and their transitive dependencies would be excluded as well. That should provide enough flexibility for everyone.

**Comment by kakawait** [comment]
> That way, you'd define the top-level artifact(s) you need to exclude and their transitive dependencies would be excluded as well

Could be strange behavior to excludes any transitive dependencies when using `<excludes><exclude>...</exclude></excludes>` pattern because (if I'm not wrong) when using it inside `<dependency></dependency>` it does not exclude transitive.

I like both @JoshuaJeme proposal IHMO.

**Comment by qzmer1104** [comment]
It's 2024 now. Not solved yet?

**Comment by philwebb** [comment]
@qzmer1104 The issue is still open and is something we'd like to fix when we can find the time.

**Comment by JoshuaJeme** [comment]
> @qzmer1104 The issue is still open and is something we'd like to fix when we can find the time.

It has been many years, can we consider the Maven ArtifactsFilter SPI solution?
Done is better then perfect. I'm very happy to contribute if I can.

**Comment by snicoll** [comment]
Anyone watching this issue, I've worked on something that gives the community more flexibility, https://github.com/snicoll/spring-boot/tree/gh-16563. 

In short, you can create your own `ArtifactsFilter` and they are invoked in order. You can also provide parameters to them as long as you respect [how Maven creates your instance](https://maven.apache.org/guides/mini/guide-configuring-plugins.html#mapping-complex-objects). 

With this change you could implement `ArtifactsFilter` in a module, add the module in the dependencies of the plugin [as shown above](https://github.com/spring-projects/spring-boot/issues/16563#issuecomment-485341142).

The commit has [an integration test](https://github.com/snicoll/spring-boot/commit/f3df6d4742b3f6d0efc55b352df81a974ed3a2a2#diff-24fcb3801a81e2f79a18c9609c2a763909650e98bcc3f657500207b4d1249de8R26-R36) that shows how to use it concretely.

This gives us several options now: custom ArtifactsFilter (new), include, exclude, exclude by groupIds, includeOptional, includeSystemScope, flagged manifest for "annotation-processor", "dependencies-starter", and "development-tool".

That's a lot and filters are applied in order. I am also wondering what `include` is all about. If you use it, only the artifacts you've defined will be included, seems a bit odd. Default filters are still applied so even if your custom filter didn't remove optional dependencies, you'd still have to set `includeOptional` to `true`.

All in all, I am wary that adding this new capability opens the door to more problems. Also, too many different options could be problematic here. 

If you're interested by this feature, please provide feedback on the above.
